crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4542,6 +4546,77 @@ impl<'db> GeneralCallableType<'db> {
             .is_some_and(|return_type| return_type.is_fully_static(db))
     }
 
+    /// Return `true` if `self` represents the exact same set of possible runtime objects as `other`.
+    pub(crate) fn is_equivalent_to(self, db: &'db dyn Db, other: Self) -> bool {
+        let self_signature = self.signature(db);
+        let other_signature = other.signature(db);
+
+        let self_parameters = self_signature.parameters();
+        let other_parameters = other_signature.parameters();
+
+        if self_parameters.len() != other_parameters.len() {
+            return false;
+        }
+
+        if self_parameters.is_gradual() || other_parameters.is_gradual() {
+            return false;
+        }
+
+        // Check equivalence relationship between two optional types. If either of them is `None`,
+        // then it is not a fully static type which means it's not equivalent either.
+        let is_equivalent = |self_type: Option<Type<'db>>, other_type: Option<Type<'db>>| match (
+            self_type, other_type,
+        ) {
+            (Some(self_type), Some(other_type)) => self_type.is_equivalent_to(db, other_type),
+            _ => false,
+        };
+
+        if !is_equivalent(self_signature.return_ty, other_signature.return_ty) {
+            return false;
+        }
+
+        for (self_parameter, other_parameter) in self_parameters.iter().zip(other_parameters) {
+            if self_parameter.name() != other_parameter.name() {
```

Comment:
I think this should not be required for positional-only parameters? Their name cannot affect whether a call succeeds or fails to bind.

Reply:
I think it's the same for variadic and keyword-variadic parameters as well because their name is not relevant in a subtype relation.

Both of the assignment is valid which means they're equivalent:

```py
class OtherCallable(Protocol):  # Supertype
    def __call__(self, *args1: int) -> Any: ...


class SelfCallable(Protocol):  # Subtype
    def __call__(self, *args2: int) -> None: ...


def _(self_callable: SelfCallable, other_callable: OtherCallable):
    a: OtherCallable = self_callable
    b: SelfCallable = other_callable
```